pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java

```diff
@@ -240,6 +243,26 @@ public void doRemoveExpiredPrimaryKeys() {
     persistWatermark(_largestSeenComparisonValue);
   }
 
+  @Override
+  public void doRemoveExpiredDeletedKeys() {
+    double threshold = _largestSeenComparisonValue - _deletedKeysTTL;
+    AtomicInteger numDeletedKeys = new AtomicInteger();
+    _primaryKeyToRecordLocationMap.forEach((primaryKey, recordLocation) -> {
```

Comment:
Looping over all entries is a very expensive operation, so we should either only filter all deleted records (using bitmap to get them), or combine it with `doRemoveExpiredPrimaryKeys()` and only loop once